src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Handler/Definitions/GoToDefinitionHandler.cs

```diff
@@ -0,0 +1,152 @@
+        {
+            var locations = ArrayBuilder<LSP.Location>.GetInstance();
+
+            var document = context.Document;
+            if (document == null)
+            {
+                return locations.ToArrayAndFree();
+            }
+
+            var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(request.Position), cancellationToken).ConfigureAwait(false);
+            var xamlGoToDefinitionService = document.Project.LanguageServices.GetService<IXamlGoToDefinitionService>();
+            if (xamlGoToDefinitionService == null)
+            {
+                return locations.ToArrayAndFree();
+            }
+
+            var definitions = await xamlGoToDefinitionService.GetDefinitionsAsync(document, position, cancellationToken).ConfigureAwait(false);
+            foreach (var definition in definitions)
+            {
+                locations.AddRange(await this.GetLocationsAsync(definition, context, cancellationToken).ConfigureAwait(false));
+            }
+
+            return locations.ToArrayAndFree();
+        }
+
+        private async Task<LSP.Location[]> GetLocationsAsync(XamlDefinition definition, RequestContext context, CancellationToken cancellationToken)
+        {
+            var locations = ArrayBuilder<LSP.Location>.GetInstance();
+
+            if (definition is XamlSourceDefinition sourceDefinition)
+            {
+                Contract.ThrowIfNull(sourceDefinition.FilePath);
+
+                var document = context.Solution?.GetDocuments(ProtocolConversions.GetUriFromFilePath(sourceDefinition.FilePath)).FirstOrDefault();
+                if (document != null)
+                {
+                    var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    var span = sourceDefinition.GetTextSpan(sourceText);
+                    if (span != null)
+                    {
+                        var location = await ProtocolConversions.TextSpanToLocationAsync(
+                                                    document,
+                                                    span.Value,
+                                                    isStale: false,
+                                                    cancellationToken).ConfigureAwait(false);
+                        locations.AddIfNotNull(location);
+                    }
+                }
+                else
+                {
+                    // Cannot find the file in solution. This is probably a file lives outside of the solution like generic.xaml
+                    // which lives in the Windows SDK folder. Try getting the SourceText from the file path.
+                    using var fileStream = new FileStream(sourceDefinition.FilePath, FileMode.Open, FileAccess.Read);
+                    var sourceText = SourceText.From(fileStream);
```

Comment:
What's the best way to handle error conditions here? For example if the file doesn't exist or cannot be opened for some reason.

Reply:
Just tried out giving it an atbitrary file path and let it throw. Don't see any error diaglog/indicator from the UI. I am thinking about let it throw if anything wrong with opening the document so we can get the exception from  Telemetry late, but yeah the downside of it would be the remaining code will not execute either (if there are more definitions). @dibarbet how does csharp handle partial success for go to definition?

Reply:
wouldn't this be a full failure - you have no location to go to?  Depending on the kind of failure we either throw or just return no definitions.

If you do not expect to hit an exception, I would leave it as throwing so you can capture telemetry on it.  If an exception here is normal and not unexpected, then it should probably be handled.  Throwing here is OK, the request queue should handle the exception and report a nfw since it isn't a mutating request.

Reply:
Thanks, David. That's what I am expecting. BTW do you know if there will be a dialog when fail to go to definition for LSP just like what we have in classic VS? Or no indication for failure is the design going forward?
![image](https://user-images.githubusercontent.com/14283750/128027656-95daf272-c6c0-47cb-afbc-73d0982ae9b6.png)